File: packages/typescript/src/session.ts
import { SessionStatus, PollSessionResponse } from './types';

type EventHandler = (...args: any[]) => void;

/**
 * Manages a single authentication relay session
 */
export class RelaySession {
  private sessionId: string;
  private status: SessionStatus = 'pending';
  private baseUrl: string;
  private apiKey: string;
  private pollInterval: number;
  private timeout: number;
  private pollTimeoutHandle?: ReturnType<typeof setTimeout>;
  private sessionTimeoutHandle?: ReturnType<typeof setTimeout>;
  private listeners: Map<string, EventHandler[]> = new Map();

  constructor(
    sessionId: string,
    baseUrl: string,
    apiKey: string,
    pollInterval: number,
    timeout: number,
  ) {
    this.sessionId = sessionId;
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.pollInterval = pollInterval;
    this.timeout = timeout;
  }

  /** Register an event handler */
  on(event: string, handler: EventHandler): this {
    if (!this.listeners.has(event)) this.listeners.set(event, []);
    this.listeners.get(event)!.push(handler);
    return this;
  }

  /** Register a one-time event handler */
  once(event: string, handler: EventHandler): this {
    const wrapped = (...args: any[]) => {
      this.off(event, wrapped);
      handler(...args);
    };
    return this.on(event, wrapped);
  }

  /** Remove an event handler */
  off(event: string, handler: EventHandler): this {
    const handlers = this.listeners.get(event);
    if (handlers) {
      this.listeners.set(event, handlers.filter(h => h !== handler));
    }
    return this;
  }

  /** Emit an event */
  private emit(event: string, ...args: any[]): void {
    const handlers = this.listeners.get(event);
    if (handlers) {
      for (const handler of [...handlers]) {
        handler(...args);
      }
    }
  }

  /** Remove all listeners */
  private removeAllListeners(): void {
    this.listeners.clear();
  }

  /** Get the current session ID */
  getSessionId(): string {
    return this.sessionId;
  }

  /** Get the current status */
  getStatus(): SessionStatus {
    return this.status;
  }

  /** Get the operator URL to share with the user */
  getOperatorUrl(): string {
    return `${this.baseUrl}/session/${this.sessionId}`;
  }

  /** Wait for the session to complete, expire, or fail */
  async wait(): Promise<Record<string, unknown> | null> {
    return new Promise((resolve, reject) => {
      this.sessionTimeoutHandle = setTimeout(() => {
        this.status = 'expired';
        this.emit('expired');
        this.stopPolling();
        reject(
          new Error(
            `AuthRelay: Session expired. The operator did not complete authentication within ${this.timeout / 1000} seconds.`,
          ),
        );
      }, this.timeout);

      const onCompleted = (credentials: Record<string, unknown>) => {
        this.cleanup();
        resolve(credentials);
      };

      const onError = (error: Error) => {
        this.cleanup();
        reject(error);
      };

      const onExpired = () => {
        this.cleanup();
        reject(
          new Error(
            `AuthRelay: Session expired. The operator did not complete authentication within ${this.timeout / 1000} seconds.`,
          ),
        );
      };

      this.once('completed', onCompleted);
      this.once('error', onError);
      this.once('expired', onExpired);

      this.startPolling();
    });
  }

  /** Abandon the session */
  async abandon(): Promise<void> {
    this.status = 'failed';
    this.stopPolling();
    this.cleanup();

    const response = await fetch(`${this.baseUrl}/api/sessions/${this.sessionId}/abandon`, {
      method: 'POST',
      headers: {
        'x-api-key': this.apiKey,
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to abandon session: ${response.statusText}`);
    }
  }

  private startPolling(): void {
    let backoffMs = 2000;
    const maxBackoffMs = 5000;

    const poll = async () => {
      try {
        const response = await fetch(`${this.baseUrl}/api/sessions/${this.sessionId}`, {
          method: 'GET',
          headers: { 'x-api-key': this.apiKey },
        });

        if (!response.ok) {
          throw new Error(`Failed to poll session: ${response.statusText}`);
        }

        const data = (await response.json()) as PollSessionResponse;
        const previousStatus = this.status;
        this.status = data.status;

        if (previousStatus !== this.status) {
          this.emit('statusChanged', this.status);
        }

        switch (data.status) {
          case 'claimed':
            if (previousStatus !== 'claimed') this.emit('claimed');
            this.pollTimeoutHandle = setTimeout(poll, backoffMs);
            backoffMs = Math.min(backoffMs + 1000, maxBackoffMs);
            break;

          case 'completed':
            if (data.credentials) {
              this.emit('completed', data.credentials);
            } else {
              this.emit('error', new Error('AuthRelay: Session completed but no credentials returned'));
            }
            this.stopPolling();
            break;

          case 'expired':
            this.emit('expired');
            this.stopPolling();
            break;

          case 'failed':
            this.emit('error', new Error(`AuthRelay: Session failed: ${data.error || 'Unknown error'}`));
            this.stopPolling();
            break;

          case 'pending':
            this.pollTimeoutHandle = setTimeout(poll, backoffMs);
            backoffMs = Math.min(backoffMs + 1000, maxBackoffMs);
            break;
        }
      } catch (error) {
        this.emit('error', error instanceof Error ? error : new Error(String(error)));
        this.stopPolling();
      }
    };

    this.pollTimeoutHandle = setTimeout(poll, 0);
  }

  private stopPolling(): void {
    if (this.pollTimeoutHandle) {
      clearTimeout(this.pollTimeoutHandle);
      this.pollTimeoutHandle = undefined;
    }
  }

  private cleanup(): void {
    this.stopPolling();
    if (this.sessionTimeoutHandle) {
      clearTimeout(this.sessionTimeoutHandle);
      this.sessionTimeoutHandle = undefined;
    }
    this.removeAllListeners();
  }
}

File: packages/python/authrelay/session.py
"""
Session management for AuthRelay
"""
import json
import time
import urllib.request
import urllib.error
from threading import Thread, Event
from typing import Optional, Dict, Any, Callable

from .types import SessionStatus, PollSessionResponse


class RelaySession:
    """Manages a single authentication relay session"""

    def __init__(
        self,
        session_id: str,
        base_url: str,
        api_key: str,
        poll_interval: int = 2000,
        timeout: int = 300_000,
    ):
        self.session_id = session_id
        self.base_url = base_url
        self.api_key = api_key
        self.poll_interval = poll_interval / 1000  # Convert to seconds
        self.timeout = timeout / 1000  # Convert to seconds
        self.status: SessionStatus = 'pending'
        self._stop_event = Event()
        self._completed_event = Event()
        self._credentials: Optional[Dict[str, Any]] = None
        self._error: Optional[Exception] = None
        self._callbacks: Dict[str, list] = {
            'claimed': [],
            'completed': [],
            'expired': [],
            'error': [],
            'statusChanged': [],
        }

    def get_session_id(self) -> str:
        """Get the current session ID"""
        return self.session_id

    def get_status(self) -> SessionStatus:
        """Get the current status"""
        return self.status

    def get_operator_url(self) -> str:
        """Get the operator URL to share with the user"""
        return f"{self.base_url}/session/{self.session_id}"

    def on(self, event: str, callback: Callable) -> None:
        """Register an event listener"""
        if event in self._callbacks:
            self._callbacks[event].append(callback)

    def off(self, event: str, callback: Callable) -> None:
        """Unregister an event listener"""
        if event in self._callbacks and callback in self._callbacks[event]:
            self._callbacks[event].remove(callback)

    def _emit(self, event: str, *args: Any) -> None:
        """Emit an event to all listeners"""
        if event in self._callbacks:
            for callback in self._callbacks[event]:
                try:
                    callback(*args)
                except Exception:
                    pass  # Silently ignore callback errors

    def wait(self, timeout: Optional[int] = None) -> Optional[Dict[str, Any]]:
        """
        Wait for the session to complete, expire, or fail.
        
        Args:
            timeout: Override timeout in seconds
            
        Returns:
            Credentials dict if successful, None otherwise
        """
        actual_timeout = timeout if timeout is not None else self.timeout
        
        # Start polling in a background thread
        poll_thread = Thread(target=self._poll_loop)
        poll_thread.daemon = True
        poll_thread.start()

        # Wait for completion or timeout
        if self._completed_event.wait(timeout=actual_timeout):
            if self._error:
                raise self._error
            return self._credentials
        else:
            # Timeout occurred
            self._stop_event.set()
            self.status = 'expired'
            self._emit('expired')
            raise TimeoutError(
                f'AuthRelay: Session expired. The operator did not complete '
                f'authentication within {actual_timeout} seconds.'
            )

    async def wait_async(self, timeout: Optional[int] = None) -> Optional[Dict[str, Any]]:
        """
        Async version of wait (requires aiohttp)
        """
        try:
            import asyncio
            from .async_client import AsyncRelaySession
            
            # Delegate to async implementation
            async_session = AsyncRelaySession(
                self.session_id,
                self.base_url,
                self.api_key,
                int(self.poll_interval * 1000),
                int(self.timeout * 1000),
            )
            return await async_session.wait(timeout)
        except ImportError:
            raise RuntimeError(
                'Async support requires aiohttp. Install with: pip install aiohttp'
            )

    def abandon(self) -> None:
        """Abandon the session"""
        self._stop_event.set()
        self.status = 'failed'

        try:
            req = urllib.request.Request(
                f"{self.base_url}/api/sessions/{self.session_id}/abandon",
                method='POST',
                headers={
                    'x-api-key': self.api_key,
                    'Content-Type': 'application/json',
                    'User-Agent': 'AuthRelay-Python-SDK/1.0.0',
                },
                data=b'',
            )
            with urllib.request.urlopen(req, timeout=10) as response:
                if response.status not in (200, 204):
                    raise Exception(f'Failed to abandon session: {response.status}')
        except urllib.error.HTTPError as e:
            raise Exception(f'Failed to abandon session: {e.reason}')
        except Exception as e:
            raise Exception(f'Failed to abandon session: {str(e)}')

    def _poll_loop(self) -> None:
        """Main polling loop (runs in background thread)"""
        backoff_ms = 2000
        max_backoff_ms = 5000
        start_time = time.time()

        while not self._stop_event.is_set():
            # Check if we've exceeded the session timeout
            elapsed = (time.time() - start_time) * 1000
            if elapsed > self.timeout * 1000:
                self.status = 'expired'
                self._emit('expired')
                self._completed_event.set()
                return

            try:
                # Poll the session status
                req = urllib.request.Request(
                    f"{self.base_url}/api/sessions/{self.session_id}",
                    method='GET',
                    headers={'x-api-key': self.api_key, 'User-Agent': 'AuthRelay-Python-SDK/1.0.0'},
                )
                with urllib.request.urlopen(req, timeout=10) as response:
                    if response.status != 200:
                        raise Exception(f'Failed to poll session: {response.status}')

                    data = json.loads(response.read().decode('utf-8'))
                    prev_status = self.status
                    self.status = data['status']

                    # Emit status changed event
                    if prev_status != self.status:
                        self._emit('statusChanged', self.status)

                    if self.status == 'claimed':
                        if prev_status != 'claimed':
                            self._emit('claimed')
                        # Continue polling with backoff
                        time.sleep(backoff_ms / 1000)
                        backoff_ms = min(backoff_ms + 1000, max_backoff_ms)

                    elif self.status == 'completed':
                        if 'credentials' in data and data['credentials']:
                            self._credentials = data['credentials']
                            self._emit('completed', self._credentials)
                        else:
                            self._error = Exception(
                                'AuthRelay: Session completed but no credentials returned'
                            )
                            self._emit('error', self._error)
                        self._completed_event.set()
                        return

                    elif self.status == 'expired':
                        self._emit('expired')
                        self._completed_event.set()
                        return

                    elif self.status == 'failed':
                        error_msg = data.get('error', 'Unknown error')
                        self._error = Exception(f'AuthRelay: Session failed: {error_msg}')
                        self._emit('error', self._error)
                        self._completed_event.set()
                        return

                    elif self.status == 'pending':
                        # Continue polling with backoff
                        time.sleep(backoff_ms / 1000)
                        backoff_ms = min(backoff_ms + 1000, max_backoff_ms)

            except urllib.error.HTTPError as e:
                self._error = Exception(f'AuthRelay: Failed to poll session: {e.reason}')
                self._emit('error', self._error)
                self._completed_event.set()
                return
            except Exception as e:
                self._error = Exception(f'AuthRelay: Polling error: {str(e)}')
                self._emit('error', self._error)
                self._completed_event.set()
                return
